lib/carbonite/migrations/v6.ex

```diff
@@ -0,0 +1,156 @@
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule Carbonite.Migrations.V6 do
+  @moduledoc false
+
+  use Ecto.Migration
+  use Carbonite.Migrations.Version
+  alias Carbonite.Migrations.V5
+
+  @type prefix :: binary()
+
+  @type up_option :: {:carbonite_prefix, prefix()}
+
+  @spec create_capture_changes_procedure(prefix()) :: :ok
+  def create_capture_changes_procedure(prefix) do
+    """
+    CREATE OR REPLACE FUNCTION #{prefix}.capture_changes() RETURNS TRIGGER AS
+    $body$
+    DECLARE
+      trigger_row #{prefix}.triggers;
+      change_row #{prefix}.changes;
+      pk_source RECORD;
+      col_name VARCHAR;
+      pk_col_val VARCHAR;
+      old_value JSONB;
+    BEGIN
+      /* load trigger config */
+      SELECT *
+        INTO trigger_row
+        FROM #{prefix}.triggers
+        WHERE table_prefix = TG_TABLE_SCHEMA AND table_name = TG_TABLE_NAME;
+
+      IF
+        (trigger_row.mode = 'ignore' AND (trigger_row.override_xact_id IS NULL OR trigger_row.override_xact_id != pg_current_xact_id())) OR
+        (trigger_row.mode = 'capture' AND trigger_row.override_xact_id = pg_current_xact_id())
+      THEN
+        RETURN NULL;
+      END IF;
+
+      /* instantiate change row */
+      change_row = ROW(
+        NEXTVAL('#{prefix}.changes_id_seq'),
+        pg_current_xact_id(),
+        LOWER(TG_OP::TEXT),
+        TG_TABLE_SCHEMA::TEXT,
+        TG_TABLE_NAME::TEXT,
+        NULL,
+        NULL,
+        '{}',
+        NULL,
+        NULL
+      );
+
+      /* build table_pk */
+      IF trigger_row.primary_key_columns != '{}' THEN
+        IF (TG_OP IN ('INSERT', 'UPDATE')) THEN
+          pk_source := NEW;
+        ELSIF (TG_OP = 'DELETE') THEN
+          pk_source := OLD;
+        END IF;
+
+        change_row.table_pk := '{}';
+
+        FOREACH col_name IN ARRAY trigger_row.primary_key_columns LOOP
+          EXECUTE 'SELECT $1.' || col_name || '::TEXT' USING pk_source INTO pk_col_val;
+          change_row.table_pk := change_row.table_pk || pk_col_val;
+        END LOOP;
+      END IF;
+
+      /* fill in changed data */
+      IF (TG_OP = 'UPDATE') THEN
+        change_row.data = to_jsonb(NEW.*) - trigger_row.excluded_columns;
+        change_row.changed_from = '{}'::JSONB;
+
+        FOR col_name, old_value
+        IN SELECT * FROM jsonb_each(to_jsonb(OLD.*) - trigger_row.excluded_columns)
+        LOOP
+          IF (change_row.data->col_name)::JSONB != old_value THEN
```

Comment:
Seems even simpler now 👍

Reply:
yep, not sure what I was thinking when I wrote this :facepalm: